Refactor: Remove Cython - Reduce tech debt

eddiebergman opened the pull request:
This PR attempts to first convert everything to Python, so a type checker can be run across everything as well as use linters properly. This also includes updating everything to pytest.

Note: This PR is unreviewable. The strategy has been to minimally change tests while refactoring to ensure tested behaviour still passes. Some notable exceptions are where functionality was removed or sampling strategy was changed and the expected values differ. Unfortunatly the formatter has rendered them also unreviewable.

**Update:** All SMAC unittests passed without code changes so it seems mostly backwards compatible.

Some notable changes:
* No more unbounded Normals
* No more quantized hyperparameters
* Sampling is 3x faster, Neighbors 3.5x faster and validation 2x faster (see scripts/benchmark_sampling.py)
* More code re-use between classes
* Pure Python, jumping to definition and docstrings should work in editors
* Defined data-structures to capture heirarchy in conditional spaces, moving the logic out of ConfigurationSpace.
* Better defined public API for ConfigurationSpace.
* Better support for custom hyperparameters, i.e. backed by Scipy Distributions (see BetaFloat/Int). Should require a lot less code to do so.
* Better defined public API for hyperparameters
* A whole load of deprecations with recommended alternatives
* Categoricals/Ordinals support arbitrary objects, e.g. no more need to specify `Category([True, False, "None"])` (Now handled in follow up #359)

TODO:
* Seems deeply nested conditions with OR's were not solved #72
* We can add a flag `Configuration(space, vector, validate=False)` to deal with #75
* Allow constants with arbitrary values #86 #95
---

* Closes #338
* Closes #336
* Closes #357
* Closes #356
* Closes #348
* Closes #344
* Closes #343
* Closes #342
* Closes #334
* Closes #330
* Closes #326 
* Closes #166
* Closes #302
* Closes #288
* Closes #237 
* Closes #270
* Closes #346
* Closes #290
* Closes #291
* Closes #293
* Closes #284
* Closes #264
* Closes #261
* Closes #262
* Closes #236
* Closes #262
* Closes #239
* Closes #206
* Closes #200
* Closes #133
* Closes #104

mfeurer reviewed (commented):
And here's a first batch of comments. I skimmed roughly 40% so far, and will probably take a bit more time...

eddiebergman commented on .github/workflows/pytest.yml:
```diff
@@ -73,136 +70,7 @@ jobs:
-
-    - name: Check for files left behind by test
-      run: |
-        before="${{ steps.status-before.outputs.BEFORE }}"
```
The main reason was actually just there was a deprecation of how input/output is handled in github actions and I couldn't get it to work with the new way. I figured the check had little value and so I discarded it.

eddiebergman commented on .github/workflows/pytest.yml:
```diff
@@ -73,136 +70,7 @@ jobs:
-        fi
-
-  # Testing with conda
-  conda-tests:
```
Yup, again, it had little value and introduced a longer wait time on getting feedback from Github actions. I don't know of a reason it wouldn't work with Conda given it's a plain Python package

eddiebergman commented on .pre-commit-config.yaml:
```diff
@@ -38,32 +38,15 @@ repos:
           - "--no-warn-return-any" # Disable this because it doesn't know about 3rd party imports
           - "--ignore-missing-imports"
           - "--show-traceback"
-      - id: mypy
```
It's there in the file from what I see, maybe just re-ordered

eddiebergman commented on pyproject.toml:
```diff
@@ -1,12 +1,12 @@
 [project]
 name = "ConfigSpace"
-version = "0.7.2"
+version = "0.8.0"
```
I'm good with that

mfeurer replied:
Actually, how about we keep this at 0.8.0 and go to 1.0 when you remove the deprecated function, introducing breaking changes?

eddiebergman commented on pyproject.toml:
```diff
@@ -112,19 +93,17 @@ exclude_lines = [
-show-source = true
-src = ["ConfigSpace", "test"]
+target-version = "py38"
+line-length = 88
```
Nothing really, just that it's migrated to `ruff` and 88 is the default for both black and ruff so I figured it would be good to just keep it at its default value

eddiebergman commented on pyproject.toml:
```diff
@@ -112,19 +93,17 @@ exclude_lines = [
+target-version = "py38"
+line-length = 88
+output-format = "full"
+src = ["src", "test"]
```
Most packaging tutorials now recommend a `src` distribution. Also prevents issues with incorrect setups where if you run `pytest` within the root directory and theres a test with `import ConfigSpace`, it's actually running the installed version rather than doing a namespaced lookup of tests.

eddiebergman commented on test/test_converters_and_test_searchspaces/test_sample_configuration_spaces.py:
```diff
@@ -27,73 +27,66 @@
+@pytest.mark.parametrize("pcs_file", pcs_files)
+def test_autosklearn_space(pcs_file: Path):
+    with warnings.catch_warnings():
+        warnings.simplefilter("ignore")
```
No, but I put a deprecation warning on using pcs read/write, citing that it's frozen and new features will not support it going forward.

Reasoning behind this is that json and yaml are more _popular_ and can be serialized from dictionaries which are easier to construct in code, while pcs files require custom strings and custom parsers, which make contributing anything new (i.e. a new HP type) much more involved. Take for example Beta parameters which don't have a pcs format but are easy to construct a json/yaml support for.

mfeurer replied:
Fair point.

eddiebergman commented on src/ConfigSpace/types.py:
```diff
@@ -0,0 +1,29 @@
+
+Number: TypeAlias = Union[int, float, np.number]
+
+npt.NDArray[np.generic]
```
That was a typo (did nothing), thanks for pointing it out and it's now removed.

eddiebergman commented on src/ConfigSpace/c_util.py:
```diff
@@ -0,0 +1,6 @@
+from ConfigSpace.util import change_hp_value, check_configuration
+
+# Backwards compatibility
+__all__ = ["check_configuration", "change_hp_value"]
```
Maybe, I was trying to minimize backwards breaking changes in general

eddiebergman commented on src/ConfigSpace/functional.py:
```diff
@@ -0,0 +1,472 @@
+    return np.concatenate([bot, top])
+
+
+def quantize_log(
```
The use of the quantization here is actually to handle integer sampling and neighborhoods, for example, quantizing a uniform (0, 1) float into an arbitrary integer range

eddiebergman commented on test/test_functional.py:
```diff
@@ -36,3 +44,180 @@ def test_arange_chunked_even():
+        (np.array([]), (0, 10), 10, np.array([])),
+    ],
+)
+def test_quantize(
```
Yeah that's true, there's no tests for it. The parameters that use it are tested though and I figured that's good enough.

eddiebergman commented on test/test_util.py:
```diff
@@ -61,567 +67,604 @@
-            "test_searchspaces",
-            "mini_autosklearn_original.pcs",
-        )
+    cs.seed(2)
```
Whoops might have been through debugging it was left in. No reason it should have been changed at the end.

eddiebergman commented on test/test_util.py:
```diff
@@ -61,567 +67,604 @@
+    all_neighbors = _test_get_one_exchange_neighbourhood(hp)
+    all_neighbors = [neighbor["a"] for neighbor in all_neighbors]
+    assert pytest.approx(np.mean(all_neighbors), abs=1e-2) == 5.44
+    assert pytest.approx(np.var(all_neighbors), abs=1e-2) == 9.14
```
So after some investigation, I found out why, the problem was to do with the test handle it called.

The first part of `def _test_get_one_exchange_neighbourhood` calculated a `num_neighbors` which for this uniform int becomes `10`.

In the second part, previously it was never passed in and the [`num_neighbors` parameter wasn't used](https://github.com/automl/ConfigSpace/blob/ce27ba26e1cf27a5b9fe242519cb622327c82420/test/test_util.py#L95-L117). In this PR, I noticed it wasn't used and must have passed it in as I thought I might have removed it accidentally.

I'm not sure what the intended behaviour is:
1. Do we remove the for loop calculating `num_neighbors`? In which case, I get something close to the previous mean and variance
2. Do we pass it in? In which case the variance is `~9` as we compute a neighbor configuration for every possible integer.

As a side note, the parameter name is misleading as the `num_neighbors` parameter to `get_one_exchange_neighbourhood` is actually the number of neighbor configs per hyperparameter, not the total number of neighbors to retrieve.

The test relies on this function:
```python
def _test_get_one_exchange_neighbourhood(hp):
    cs = ConfigurationSpace()
    num_neighbors = 0
    if not isinstance(hp, list):
        hp = [hp]

    for hp_ in hp:
        cs.add(hp_)
        if np.isinf(hp_.get_num_neighbors()):
            num_neighbors += 4
        else:
            num_neighbors += hp_.get_num_neighbors()

    cs.seed(1)
    config = cs.get_default_configuration()
    all_neighbors = []
    for i in range(100):
        neighborhood = get_one_exchange_neighbourhood(
            config,
            i,
			# Before, num_neighbors was set as blank, which defaulted to 4
            num_neighbors=4,
			# The test after updating had it set to the above `num_neighbors`
			num_neighbors=num_neighbors
        )

```

mfeurer replied:
> Do we remove the for loop calculating num_neighbors? In which case, I get something close to the previous mean and variance

I guess it's fine to keep the loop. Thanks for checking what caused the changed numbers.

> As a side note, the parameter name is misleading as the num_neighbors parameter to get_one_exchange_neighbourhood is actually the number of neighbor configs per hyperparameter, not the total number of neighbors to retrieve.

Would it be possible to change this?

eddiebergman commented on test/test_util.py:
```diff
@@ -61,567 +67,604 @@
+    all_neighbors = [neighbor["a"] for neighbor in all_neighbors]
+    assert hp.default_value == 3
+
+    assert pytest.approx(np.mean(all_neighbors), abs=1e-2) == 5.77
```
See above

eddiebergman commented on test/test_util.py:
```diff
@@ -61,567 +67,604 @@
+    assert hp.default_value == 3
+
+    assert pytest.approx(np.mean(all_neighbors), abs=1e-2) == 5.77
+    assert pytest.approx(np.var(all_neighbors), abs=1e-2) == 8.39
```
And above

mfeurer reviewed (commented):
Part 2: I finished looking at all the tests. Looks really great, I've only got a few questions this time.

eddiebergman commented on test/test_hyperparameters.py:
```diff
@@ -48,1909 +46,1865 @@
+    wrong_shape_2 = np.array([3, 5, 7]).reshape(1, -1)
+    wrong_shape_3 = np.array([3, 5, 7]).reshape(-1, 1)
+
+    assert c1.pdf_values(point_1)[0] == pytest.approx(0.7559999999999997)
```
Ay yes, this I didn't know how to handle properly.

The previous implementation returned the pdf in the value space (even when considering vectorized values).
https://github.com/automl/ConfigSpace/blob/ce27ba26e1cf27a5b9fe242519cb622327c82420/ConfigSpace/hyperparameters/beta_float.pyx#L189-L213

You can see from the bottom of that function it normalizes with respect to `self.upper` and `self.lower`, numbers in value space. In this case reducing it by a factor of 10 (and not 5 as you mentioned).

The newer implementation essentially has hyperparameters being backed by some distribution in vectorized space, that have no idea about what their vectorized format means in value space. The new values reflect this in they not normalized by the lower and upper bound in value space. I figured this was okay as `pdf` only has a relative meaning when compared to other pdfs, and they all no longer have this value space normalization.

Here's the relevant snippet from the code. I did originally use to pass in the `_pdf_norm` to rescale things but i figured it was better to have the pdf refer to their pdf in vectorized space and be agnostic to value space.

```python
@dataclass
class ScipyContinuousDistribution(Distribution):
    rv: rv_continuous_frozen

    lower_vectorized: f64
    upper_vectorized: f64
    _max_density: float
    _pdf_norm: float = 1

    def sample_vector(
        self,
        n: int,
        *,
        seed: RandomState | None = None,
    ) -> Array[f64]:
        return self.rv.rvs(size=n, random_state=seed).astype(f64)  # type: ignore

    def max_density(self) -> float:
        return float(self._max_density)

    def pdf_vector(self, vector: Array[f64]) -> Array[f64]:
        pdf = self.rv.pdf(vector) / self._pdf_norm

        # By definition, we don't allow NaNs in the pdf
        return np.nan_to_num(pdf, nan=0)  # type: ignore
```

As a side beneift, this means its first of all easier to implement new hps as you do not need to consider the interplay of pdf w.r.t. vectorized and value space and secondly if we just rely on the `pdf` returned by the underlying disitribution, things become a lot more uniform in terms of implementation.

eddiebergman commented on test/test_configuration_space.py:
```diff
@@ -578,686 +568,761 @@ def test_check_configuration(self):
+    assert cs["algo2_subspace:algo2_param1"].default_value == "Y"
+
+
+def test_acts_as_mapping_2():
```
Was an issue where two tests in the same file had the same name but were part of test classes. Making it `pytest` and removing the classes caused it to fail and I think this was me being lazy.

Renamed to `test_configuration_space_acts_as_mapping` and `test_configuration_acts_as_mapping`

sonovice commented:
For everyone else being stuck due to type errors etc:
You can pull in the current state of this PR to your poetry env with
```
poetry add git+https://github.com/automl/ConfigSpace.git@refs/pull/346/merge
```

Might help in your case.

mfeurer commented on src/ConfigSpace/_condition_tree.py:
```diff
@@ -0,0 +1,698 @@
+
+@dataclass
+class DAG:
+    # All relevant information is kept in these fields, rest is cached information
```
Could you please make this comment more specific? Right now it is not clear to which variables you mean by "these fields". I would have expected "keep track of conditions" is also relevant, because the conditions also the vertices of the graph.

eddiebergman replied:
Yup, I will have to take time for all the documentation!

mfeurer commented on test/test_util.py:
```diff
@@ -175,6 +176,7 @@ def test_random_neighborhood_int():
     assert pytest.approx(np.mean(all_neighbors), abs=1e-2) == 5.77
     assert pytest.approx(np.var(all_neighbors), abs=1e-2) == 8.39
 
+    return
```
Should this be here?

eddiebergman replied:
Nope, good spot

mfeurer commented on test/test_util.py:
```diff
@@ -100,9 +100,10 @@ def _test_get_one_exchange_neighbourhood(hp):
+            num_neighbors=4,
         )
         ns = list(neighborhood)
+        print(config["a"], ns)
```
Should this be here?

eddiebergman replied:
Nope, good spot!

eddiebergman commented on src/ConfigSpace/hyperparameters/categorical.py:
```diff
@@ -0,0 +1,266 @@
+class CategoricalHyperparameter(Hyperparameter[Any, Any]):
+    ORDERABLE: ClassVar[bool] = False
+
+    choices: Sequence[Any]
```
Yup, sorry, I should go through and document these properly!

eddiebergman commented on src/ConfigSpace/hyperparameters/_hp_components.py:
```diff
@@ -0,0 +1,337 @@
+from __future__ import annotations
```
The reason for the abstraction of things in this file is that a lot of methods for the hyperparameters were essentially doing the same things, i.e. sampling from some vectorized distribution and then transforming to and from.

The original thought was that this would be one of the cythonized files, i.e. allow the hyperparameter files to be seen in editors and see their definitions and workings, however this also ended up with lifting all behavior to the base Hyperparameter class. This helps for a few reasons, namely:
1. A single source for behaviour of hyperparameters, reducing duplication/inconsistencies.
2. Implementing hyperparameters is mostly just passing the different components described above to the base `Hyperparameter` class at this point. You can see this from the individual hyperparameter classes that exist.
3. Any optimizations done for the underlying shared components can apply to all hyperparmaters, for example neighbor generation or sampling.

It's not perfect of course but really it's trying to make hyperparameters rely more on compositional structure rather than hierarchical behaviors.

eddiebergman commented on src/ConfigSpace/hyperparameters/_distributions.py:
```diff
@@ -0,0 +1,947 @@
+from __future__ import annotations
```
See above

eddiebergman commented on src/ConfigSpace/hyperparameters/ordinal.py:
```diff
@@ -0,0 +1,122 @@
+class OrdinalHyperparameter(Hyperparameter[Any, Any]):
+    ORDERABLE: ClassVar[bool] = True
+
+    sequence: tuple[Any, ...]
```
See above

eddiebergman commented:
@mfeurer I converted all the documentation pages to use `mkdocs-material` format, i.e. markdown documentation.
The primary reason is I know how to build docs/version and deploy them. It's also a much less finicky syntax, i.e. I never get a sphinx link correct.

I've documented most of the major modules at this point but have yet to set up the actual doc generation process. What I think would be the most useful for you to review at this point would be the pages under [`docs/reference`](https://github.com/automl/ConfigSpace/tree/typing-cython-3/docs/reference).

I mostly left the documentation the same but I've added a clearer introduction to [hyperparameters](https://github.com/automl/ConfigSpace/blob/typing-cython-3/docs/reference/hyperparameters.md?plain=1)

Due to a no-longer-updated take on PCS format, I've now focused serialization on `json()` and `yaml()` which made this page change: https://github.com/automl/ConfigSpace/blob/typing-cython-3/docs/reference/serialization.md?plain=1
The encoder/decoder part is to satisfy this issue #357. It also enables us to allow people to create their own import and export formats as required for them.

The rest of the docs are mostly the same.

---

I still have left to do API documentation pass over `ConfigurationSpace` and `Configuration` as well as the `utils.py` file.

---

After that I need to setup `mkdocs` and fix any potential typoed links and such. The look and layout will be mostly similar to [`amltk` docs](https://automl.github.io/amltk/latest/).

Sorry for migration away from sphinx, but I really just don't like writing in it and it's hard to explain the nuances to anyone new.

sarah-segel commented:
FYI: I tested running DeepCave with ConfigSpace 0.8.0 and ran into the following problems when executing :

1) I run into an error when calling
`from ConfigSpace.c_util import change_hp_value, check_forbidden
`
2) It seems like the default values in my configspace.json cause a problem when calling
`
from ConfigSpace.read_and_write import json as cs_json
`
`
cs_json.read(self.configspace_fn.read_text())
`
where I get:
`__init__() got an unexpected keyword argument 'default'
`
3) When calling hp._transform_scalar(value) I get 
`AttributeError: 'UniformFloatHyperparameter' object has no attribute '_transform_scalar'`

eddiebergman commented:
@mfeurer, I did a major push on getting the documentation updated to `mkdocs`. You can view it locally by cloning this branch, `pip install -e "[dev]"` and running `mkdocs serve`. Likely another push should get this to the finish line.

@sarah-segel thanks for the report back, I will look into them with the next big push.

eddiebergman commented:
@sarah-segel should have been fixed now, can you try again?

mfeurer commented on test/test_conditions.py:
```diff
@@ -49,6 +49,10 @@
 from ConfigSpace.hyperparameters.hyperparameter import Hyperparameter
 
 
+def some_new_function():
```
What do we need this one for?

eddiebergman replied:
Good spot, removed

mfeurer commented on src/ConfigSpace/functional.py:
```diff
@@ -0,0 +1,414 @@
+
+
+def center_range(
+    center: int,
```
What happens if the difference from the lower to the center is smaller than from the center to higher? Is this an issue?

eddiebergman replied:
Nope, `roundrobin()` will start just giving those from the `above_center` only after it can no longer yield equally between `below_center` and `above_center`

mfeurer commented on src/ConfigSpace/hyperparameters/hyperparameter.py:
```diff
@@ -0,0 +1,780 @@
+        """Initialize a hyperparameter.
+
+        Args:
+            name:
```
Is it necessary to duplicate this information here?

eddiebergman replied:
Not entirely but I don't have it a different way. This is documenting parameters, while the other ones are documenting attributes. I would say it's necessary as it shows up in editors as help information

mfeurer commented on src/ConfigSpace/hyperparameters/hyperparameter.py:
```diff
@@ -0,0 +1,780 @@
+        )
+
+    def pdf_vector(self, vector: Array[f64]) -> Array[f64]:
+        """Get the probability density of an array vectorized values.
```
```suggestion
        """Get the probability density of an array of vectorized values.
```

mfeurer commented on src/ConfigSpace/hyperparameters/hyperparameter.py:
```diff
@@ -0,0 +1,780 @@
+            The probability density of the values. Where values are not legal,
+            the probability density is zero.
+        """
+        # TODO(eddiebergman): Backwards compatible restriction, why this restriction?
```
This would be great to know ;)

eddiebergman replied:
Will leave it in now unless it's a problem at some point

mfeurer commented on src/ConfigSpace/hyperparameters/hp_components.py:
```diff
@@ -0,0 +1,509 @@
+        return vector * self._size + _l  # type: ignore
+
+    def _unsafe_to_value(self, vector: Array[f64]) -> Array[f64]:
+        # NOTE: Unsafe as it does not check boundaries, clip or integer'ness
```
Could you please add this comment to the function above, too?

mfeurer reviewed (commented):
Only ten files to go

mfeurer commented on src/ConfigSpace/_condition_tree.py:
```diff
@@ -0,0 +1,909 @@
+    if choice != A:
+        then A:x, A:y, A:z are inactive
+
+    # NOTE: I'm fairly sure this assumption still leads to a diamond OR condition
```
Should this maybe rather be a TODO?

mfeurer commented on src/ConfigSpace/hyperparameters/distributions.py:
```diff
@@ -0,0 +1,1301 @@
+    """Create a neighborhood of `n` neighbors around `vector` with a normal distribution.
+
+    The neighborhood is created by sampling from a normal distribution centered around
+    `vector` with a standard deviation of `std`. The samples are then quantized to the
```
Do you rather mean truncated?

mfeurer replied:
Also, there are no bins here, right?

eddiebergman replied:
Yup, good spot, cleaned up the documentation for this.

mfeurer reviewed (commented):
I finished looking at all files :) Now I need to look at your answer to my previous questions...

eddiebergman commented on .github/workflows/docs.yml:
```diff
@@ -1,72 +1,27 @@
+      - name: "Install dependancies"
+        run: python -m pip install -e ".[dev]"
+      - name: "Build Docs"
+        run: mkdocs build --clean --strict
```
Yup, we moved away from Sphinx and I didn't know how to automate doc deployments given this setup. I will document how to do it in manual way in `CONTRIBUTING.md`

eddiebergman commented on docs/reference/hyperparameters.md:
```diff
@@ -0,0 +1,546 @@
+    ORDERABLE: ClassVar[bool] = True  # Let ConfigSpace know there is an order to the values
+
+    alpha: float
+    """Some docstring decsription of this attribute."""
```
Nope, it's alright right here. It puts the documentation right with the declaration.

eddiebergman commented:
Will be making a release once these tests pass :)
